File: scripts/fb6fbaba/local/kcore.cpp
#include "graph.h"
#include "kcore.h"

#include <algorithm>
#include <cassert>
#include <filesystem>
#include <format>
#include <fstream>
#include <iostream>
#include <map>
#include <set>
#include <vector>

// TODO: refactor to using vectors instead of sets whenever possible

Vertex update_uppers(Graph *g, Vertex cur, std::set<Vertex> &removed, std::vector<Vertex> &uppers)
{
    // the new estimate is the maximum k s.t.
    // number of neighbors(v) with core value >= k is more than k

    Vertex degree = get_degree(g, cur);
    Vertex *nbrs = neighbors(g, cur);

    // nums[k-1] = # neighbors with core (estimate) k
    std::vector<Vertex> nums(uppers[cur]);
    for (Vertex i = 0; i < degree; ++i)
    {
        if (!removed.contains(nbrs[i]))
        {
            assert(uppers[nbrs[i]] > 0);
            ++nums[std::min(uppers[cur], uppers[nbrs[i]]) - 1];
        }
    }

    Vertex k, accum = 0;
    for (k = uppers[cur]; k > 1; --k)
        if ((accum += nums[k - 1]) >= k)
            break;

    return k;
}

void update_lowers(std::map<Vertex, std::set<Vertex>> &graph, Vertex max_degree, std::set<Vertex> &candidates,
                   std::vector<Vertex> &lower)
{
    std::vector<std::set<Vertex>> queues(max_degree + 1);
    std::map<Vertex, Vertex> sub_degrees;
    for (Vertex v : candidates)
    {
        sub_degrees[v] = graph[v].size();
        queues[sub_degrees[v]].insert(v);
    }

    for (Vertex k = 1; k <= max_degree; ++k)
    {
        while (!queues[k].empty())
        {
            Vertex u = *queues[k].begin();
            queues[k].erase(u);
            lower[u] = k;

            for (Vertex v : graph[u])
            {
                if (candidates.contains(v) && sub_degrees[v] > k)
                {
                    queues[sub_degrees[v]].erase(v);
                    queues[sub_degrees[v] - 1].insert(v);
                    --sub_degrees[v];
                }
            }
        }
    }
}

std::set<Vertex> get_comm(Graph *graph, Vertex q, Vertex k)
{
    std::vector<Vertex> lowers(graph->n_nodes), uppers(graph->n_nodes);
    std::vector<Vertex> counts(graph->n_nodes);
    std::set<Vertex> need_update;
    std::set<Vertex> candidates, removed, frontier;

    std::map<Vertex, std::set<Vertex>> sub_graph;
    Vertex max_degree = 0;

    candidates.insert(q);
    need_update.insert(q);
    uppers[q] = get_degree(graph, q);

    Vertex *q_nbrs = neighbors(graph, q);
    for (Vertex i = 0; i < get_degree(graph, q); ++i)
        frontier.insert(q_nbrs[i]);

    while (!need_update.empty())
    {
        if (uppers[q] < k)
            return {};

        // stage 0: expand the frontier
        for (Vertex cur : frontier)
            uppers[cur] = get_degree(graph, cur);

        // step 1: iterate the lower bounds
        update_lowers(sub_graph, max_degree, candidates, lowers);

        // stage 2: iterate the upper bounds
        std::set<Vertex> need_update_next;
        for (Vertex cur : need_update)
        {
            Vertex degree = get_degree(graph, cur);
            Vertex *nbrs = neighbors(graph, cur);

            Vertex c_old = uppers[cur];
            Vertex c_new = update_uppers(graph, cur, removed, uppers);
            uppers[cur] = c_new;

            counts[cur] = 0;
            for (Vertex i = 0; i < degree; ++i)
            {
                Vertex u = nbrs[i];
                if (!candidates.contains(u))
                    continue;

                if (uppers[cur] < uppers[u] && uppers[u] <= c_old)
                    --counts[u];
                if (counts[u] < uppers[u])
                    need_update_next.insert(u);
                if (uppers[cur] <= uppers[u])
                    ++counts[cur];
            }
        }
        std::swap(need_update_next, need_update);

        // step 3: figure out the updates for next iteration
        std::set<Vertex> frontier_next;
        for (Vertex cur : frontier)
        {
            candidates.insert(cur);
            need_update.insert(cur);

            Vertex degree = get_degree(graph, cur);
            Vertex *nbrs = neighbors(graph, cur);
            for (Vertex i = 0; i < degree; ++i)
            {
                Vertex u = nbrs[i];

                if (get_degree(graph, u) < std::min(k, lowers[q]))
                {
                    // TODO: counts and updates
                    removed.insert(u);
                    continue;
                }

                if (!candidates.contains(u) && !removed.contains(u))
                {
                    frontier_next.insert(u);
                }
                else
                {
                    sub_graph[cur].insert(u);
                    sub_graph[u].insert(cur);

                    max_degree = std::max(max_degree, sub_graph[cur].size());
                    max_degree = std::max(max_degree, sub_graph[u].size());
                }
            }
        }
        std::swap(frontier_next, frontier);

        // remove all nodes with UB[cur] < LB[q] out of candidates
        std::erase_if(candidates, [&](Vertex v) {
            if (uppers[v] >= lowers[q])
                return false;

            removed.insert(v);
            for (Vertex u : sub_graph[v])
                sub_graph[u].erase(v);
            sub_graph.erase(v);

            return true;
        });
    }

    std::set<Vertex> community;
    for (Vertex cur : candidates)
    {
        std::cout << lowers[cur] << "<=Core(" << cur << ")<=" << uppers[cur] << "\n";
        assert(lowers[cur] == uppers[cur]);
        if (lowers[cur] >= lowers[q])
            community.insert(cur);
    }

    std::cout << "Query " << q << " " << lowers[q] << "<=Core(q)<=" << uppers[q] << " Considered "
              << removed.size() + candidates.size() << " Outputted " << community.size() << "\n";

    return community;
}

void write_comm(char *maindir, Vertex q, Vertex k, std::set<Vertex> &comm)
{
    std::filesystem::path dir = std::filesystem::path(maindir) / std::to_string(q);
    std::filesystem::path path = dir / std::format("kcore_{}.txt", k);
    std::filesystem::create_directories(dir);

    std::ofstream file(path.string());
    for (Vertex v : comm)
        file << v << "\n";
}

File: scripts/fb6fbaba/local/kcore.py
import itertools as it
from pathlib import Path

import networkx as nx


def get_comm(g: nx.Graph, q: int, query_k: int) -> set[int]:
    def update_lowers(subg: nx.Graph, lowers, finished):
        # subg = nx.induced_subgraph(subg, set(subg.nodes).difference(frntr))
        degrees = dict(subg.degree)

        max_degree = max(degrees.values())
        queues = [set() for _ in range(max_degree + 1)]
        for v, degree in subg.degree:
            queues[degree].add(v)

        update = False
        for k in range(max_degree + 1):
            while queues[k]:
                v = queues[k].pop()
                if lowers[v] != k:
                    lowers[v] = k

                    update = v not in finished
                    if k >= query_k:
                        finished.add(v)

                for u in subg.neighbors(v):
                    if degrees[u] > k:
                        queues[degrees[u]].remove(u)
                        queues[degrees[u] - 1].add(u)
                        degrees[u] -= 1
        return update

    def update_uppers(subg: nx.Graph, uppers, finished):
        def local_core(v):
            nums = [0] * (uppers[v] + 1)
            for u in g.neighbors(v):
                nums[min(uppers[v], uppers.get(u, g.degree[u]))] += 1
                # nums[min(uppers[v], uppers[u])] += 1

            accum = 0
            for k in reversed(range(len(nums))):
                if (accum := accum + nums[k]) >= k:
                    return k

        update = False
        for v in subg.nodes:
            c_new = local_core(v)
            if c_new != uppers[v]:
                uppers[v] = c_new
                update = v not in finished

        return update

    def update_subgraph(g: nx.Graph, subg: nx.Graph, lowers, uppers):
        num_nodes = subg.number_of_nodes()
        update = False
        visited = {q}
        queue = [q]

        while queue:
            v = queue.pop()

            for u in g.neighbors(v):
                if u in visited:
                    continue

                if u not in lowers or u not in uppers:
                    lowers[u] = 0
                    uppers[u] = g.degree(u)

                if uppers[u] >= max(lowers[q], query_k):
                    visited.add(u)

                    if u not in subg.nodes:
                        update = True
                    else:
                        queue.append(u)

        # print(f"{subg.number_of_nodes()=} {len(visited)=}")

        new_subg = nx.induced_subgraph(g, visited)
        update |= len(visited) != num_nodes
        return new_subg, update

    lowers, uppers = dict(), dict()
    finished = set()

    # initialize with the 1-hop neighborhood of q
    subg = nx.induced_subgraph(g, list(g.neighbors(q)) + [q])
    lowers[q] = 0
    uppers[q] = g.degree[q]
    for u in g.neighbors(q):
        lowers[u] = 0
        uppers[u] = g.degree[u]

    # subg = g.copy()
    # lowers = {v: 0 for v in g.nodes}
    # uppers = {v: g.degree[v] for v in g.nodes}
    # frntr = set()

    all_visited = set(subg.nodes).copy()

    update = True
    while update and uppers[q] >= query_k:
        print(f"{lowers[q]=}, {uppers[q]=} {subg.number_of_nodes()=}")

        update_lower = update_lowers(subg, lowers, finished)
        update_upper = update_uppers(subg, uppers, finished)

        subg, update_subg = update_subgraph(g, subg, lowers, uppers)
        update = update_lower | update_upper | update_subg

        all_visited = all_visited.union(set(subg.nodes))

    # lowers = {v: 0 for v in g.nodes}
    # update_lowers(g, lowers)
    # uppers = {v: g.degree[v] for v in g.nodes}
    # for _ in it.count():
    #     if not update_uppers(g, uppers):
    #         break
    # for v in g.nodes:
    #     assert lowers[v] == uppers[v]

    print(f"{lowers[q]=}, {uppers[q]=}")
    print(len(all_visited), len(subg.nodes))
    return set(subg.nodes)


def main():
    g: nx.Graph = nx.read_edgelist("test/network.tsv", nodetype=int)
    with open("test/nodes.txt") as f:
        lines = [map(int, line.strip().split()) for line in f.readlines()]

    for q, k in lines:
        print(f"Getting comm for {q=} {k=}")
        cmty = get_comm(g, q, k)
        path = Path(f"test/{q}/kcore_{k}.txt")
        path.parent.mkdir(parents=True, exist_ok=True)
        with path.open("w") as f:
            f.write("\n".join(map(str, cmty)))

        break


if __name__ == "__main__":
    main()
